src/decoding.rs

```diff
@@ -259,7 +259,38 @@ pub fn decode<T: DeserializeOwned>(
     key: &DecodingKey,
     validation: &Validation,
 ) -> Result<TokenData<T>> {
-    match verify_signature(token, key, validation) {
+    decode_bytes(token.as_bytes(), key, validation)
+}
+
+/// Decode and validate a JWT
+///
+/// If the token or its signature is invalid or the claims fail validation, it will return an error.
+///
+/// This differs from decode() in the case that you only have bytes. By decoding as bytes you can
+/// avoid taking a pass over your bytes to validate them as a utf-8 string. Since the decoding and
+/// validation is all done in terms of bytes, the &str step is unnecessary.
+/// If you already have a &str, decode is more convenient. If you have bytes, consider using this.
+///
+/// ```rust
+/// use serde::{Deserialize, Serialize};
+/// use jsonwebtoken::{decode_bytes, DecodingKey, Validation, Algorithm};
+///
+/// #[derive(Debug, Serialize, Deserialize)]
+/// struct Claims {
+///    sub: String,
+///    company: String
+/// }
+///
+/// let token = b"a.jwt.token";
+/// // Claims is a struct that implements Deserialize
+/// let token_message = decode_bytes::<Claims>(token, &DecodingKey::from_secret("secret".as_ref()), &Validation::new(Algorithm::HS256));
+/// ```
+pub fn decode_bytes<T: DeserializeOwned>(
```

Comment:
any reason to not change `decode` to take `token: impl AsRef<[u8]>` and have a single fn?
Same question for decode_header

Reply:
Honestly, I think that would be better. It might be a minor breaking change for your users who use clippy. I was trying to avoid any potential for an api break, but that may not be the right thing to optimize for.
If you're willing to permit it, I'd be happy to see how taking token as AsRef feels!

Reply:
I think it looks good, I'm going to wait a bit to see if there are more breaking changes to add for a new major version though

Reply:
Yep, seems like there are some other reasonable candidates for breaking changes in your issue queue.

Thanks for reviewing!